### ISSUE TITLE
export YOLO11 onnx with NMS 

### ISSUE THREAD

**Issue by Peterande** [ultralytics/ultralytics#16732]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I used the following code to export my model with ONNX and enabled  NMS:

`path = model.export(format="onnx", nms=True)`

However, the exported ONNX model doesn't seem to include NMS. Could you please let me know how to properly add NMS to the exported ONNX model? Additionally, what are the confidence threshold and IoU threshold values used for NMS in the official reported mAP?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Peterande, thank you for reaching out! We appreciate your interest in Ultralytics 🚀. This is an automated response, but rest assured, an Ultralytics engineer will assist you soon.

In the meantime, I recommend checking out our [Docs](https://docs.ultralytics.com) for guidance on exporting models and integrating NMS (Non-Maximum Suppression).

To help us better understand the issue, could you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/)? This will aid in debugging.

For any further queries or detailed discussions, you can join our community:

- Real-time chat on [Discord](https://ultralytics.com/discord) 🎧
- In-depth discussions on [Discourse](https://community.ultralytics.com)
- Share experiences on our [Subreddit](https://reddit.com/r/ultralytics)

## Upgrade

Make sure to upgrade to the latest `ultralytics` package in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) using:

```bash
pip install -U ultralytics
```

## Environments

You can execute YOLO in various environments with dependencies preinstalled:

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If you see a green badge, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. These tests ensure correct functionality across various platforms.

Stay tuned, and thank you for your patience! 😊

**Comment by Y-T-G** [comment]
> Additionally, what are the confidence threshold and IoU threshold values used for NMS in the official reported mAP?

IoU threshold is 0.6. During mAP calculation, a very low confidence threshold is applied (0.001) due to how mAP is calculated.

You can check the docs for default arguments used:
https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation

**Comment by Skillnoob** [comment]
If you read the docs, it says that the nms argument is only availabe for coreml.
https://docs.ultralytics.com/modes/export/#export-formats

**Comment by Peterande** [comment]
Thank you for your quick reply.

**Comment by YuriGao** [comment]
@glenn-jocher Hello, glenn. I still wonder why we coundn't export format ONNX with NMS suppot? According to ONNX Opset 11, there is NonMaxSuppression Op support. Wish for your reply.

**Comment by Skillnoob** [comment]
Ultralytics does not have built in support for adding nms to the onnx models when exporting. You can open a feature request issue and continue from there.

**Comment by Peterande** [comment]
> @glenn-jocher Hello, glenn. I still wonder why we coundn't export format ONNX with NMS suppot? According to ONNX Opset 11, there is NonMaxSuppression Op support. Wish for your reply.

Try https://github.com/Peterande/D-FINE/blob/master/tools/deployment/export_yolo_w_nms

**Comment by glenn-jocher** [comment]
@Peterande thank you for the suggestion. We currently don't support exporting ONNX with NMS directly, but you can explore custom solutions like the one you mentioned.

**Comment by Egorundel** [comment]
> > @glenn-jocher Hello, glenn. I still wonder why we coundn't export format ONNX with NMS suppot? According to ONNX Opset 11, there is NonMaxSuppression Op support. Wish for your reply.
> 
> Try https://github.com/Peterande/D-FINE/blob/master/tools/deployment/export_yolo_w_nms

Has anyone checked? Does it work?

**Comment by glenn-jocher** [comment]
@Egorundel i recommend testing the solution in your environment to verify its compatibility. If you encounter issues, feel free to ask for further assistance.

**Comment by salmanmaq** [comment]
I patched together some stuff and came up with this working version: https://gist.github.com/salmanmaq/6c444085184dc6129883421b954b12bb

Best reference is https://vilsonrodrigues.medium.com/add-non-maximum-suppression-nms-to-object-detection-model-using-onnx-7639a698cf05

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution and references. It's great to see community contributions helping to address the NMS integration with ONNX models. If you have further questions or need assistance, feel free to ask.

**Comment by fitoule** [comment]
> I patched together some stuff and came up with this working version: https://gist.github.com/salmanmaq/6c444085184dc6129883421b954b12bb
> 
> Best reference is https://vilsonrodrigues.medium.com/add-non-maximum-suppression-nms-to-object-detection-model-using-onnx-7639a698cf05

Hi @salmanmaq  I've just tried your code whithout changing anything and it does not work  :
![bus](https://github.com/user-attachments/assets/5b207f62-9b8b-48ba-aa89-9a0d1969510a)
my test command was  `yolo predict model="/media/yolo11l_final.onnx" source='https://ultralytics.com/images/bus.jpg'`

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #18185](https://github.com/ultralytics/ultralytics/pull/18185) 🎉! Here's a summary of the key changes introduced in this PR:

### Key Updates:
- **Improved Argument Validation**: A new `validate_args` function has been added to ensure only supported arguments are passed during export, reducing the likelihood of errors.
- **Updated Export Format Configurations**: Specific configurations now support features like `half`, `int8`, and `batch`, improving clarity and usability.
- **Refined Workflows**: The export and benchmarking processes now leverage enhanced validations, making the process more robust and predictable.
- **Version Bump**: The library version has been updated from `8.3.52` to `8.3.53`.

These updates aim to enhance the export functionality, reduce errors, and improve overall usability. If this addresses the issue you were experiencing, we encourage you to test it out!

### How to Test:
1. Install the latest changes directly from the `main` branch:
   ```bash
   pip install git+https://github.com/ultralytics/ultralytics.git@main
   ```
2. OR wait until the next official PyPI release to test the updated version.

### Request for Feedback:
Please let us know if the changes in this PR resolve your issue. If the problem persists or you encounter any new challenges, feel free to provide additional feedback. 

Thank you for reporting this issue 🙏! Your input is invaluable in helping us improve. 🚀
